services/web-app/pages/_middleware.js

```diff
@@ -32,12 +32,26 @@ function exitWith404(req) {
 }
 
 export async function middleware(req) {
+  const protocol =
+    getRunMode === RunMode.Prod || process.env.RUNNING_SECURELY === '1' ? 'https' : 'http'
+
+  // log page request
+  fetch(`${protocol}://localhost:${req.nextUrl.port}/api/log-message`, {
+    method: 'POST',
+    body: JSON.stringify({
+      logMessage: `[${new Date().toISOString()}] "${req.method} ${req.nextUrl.pathname}" "${
+        req.ua.ua
+      }" "${req.ip}"`
+    }),
+    headers: {
+      'Content-Type': 'application/json'
+    }
+  })
+
```

Comment:
This middleware function is getting a bit lengthy and does a few different things. Would you mind refactoring it to delegate the individual pieces of the middleware to their own functions which are called by `middleware()`? Here's a starting point that I had locally when I tried to do the auth stuff. You could add another function to the `middleware` function to perform the logging.

```js
/*
 * Copyright IBM Corp. 2022, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { getRunMode, RunMode } from '@carbon-platform/api/runtime'
import cookie from 'cookie'
import { NextResponse } from 'next/server'

import { isValidIbmEmail } from '@/utils/string'

class NotAuthorizedError extends Error {}

/**
 * Creates request options for retrieveUser api call including headers and agent if necessary
 *
 * @param {NextRequest} req getServerSideProps request object
 * @returns {RequestInit} request options
 */
function getRequestOptions(req) {
  const sessionCookie = req.cookies['connect.sid']
  const cookies = sessionCookie ? cookie.serialize('connect.sid', sessionCookie) : ''
  return {
    headers: { cookie: cookies }
  }
}

function exitWith404(req) {
  const url = req.nextUrl.clone()
  url.pathname = '/404'
  return NextResponse.rewrite(url)
}

async function applyAuthMiddleware(req) {
  if (req.nextUrl.pathname === '/404' || req.nextUrl.pathname.startsWith('/api/')) {
    return
  }

  const protocol =
    getRunMode === RunMode.Prod || process.env.RUNNING_SECURELY === '1' ? 'https' : 'http'
  const userResponse = await fetch(
    `${protocol}://localhost:${req.nextUrl.port}/api/user`,
    getRequestOptions(req)
  )

  // Guard - non-200 user api response
  if (!userResponse.ok) {
    throw new NotAuthorizedError()
  }

  const user = await userResponse.json()

  // Guard - no valid user
  if (!user?.email) {
    throw new NotAuthorizedError()
  }

  // Guard - not a valid IBMer
  if (!isValidIbmEmail(user.email)) {
    throw new NotAuthorizedError()
  }
}

export async function middleware(req) {
  try {
    await applyAuthMiddleware(req)
  } catch (err) {
    return exitWith404(req)
  }

  return NextResponse.next()
}
```

Reply:
Thanks for the file! incorporated these changes and added call `logIncomingRequest()` to `middleware`